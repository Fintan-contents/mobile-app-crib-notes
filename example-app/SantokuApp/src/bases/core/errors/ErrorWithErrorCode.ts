export type ErrorWithErrorCode = Error & {errorCode: string};
export const isErrorWithErrorCode = (error: unknown): error is ErrorWithErrorCode =>
  error != null && error instanceof Error && 'errorCode' in error;
