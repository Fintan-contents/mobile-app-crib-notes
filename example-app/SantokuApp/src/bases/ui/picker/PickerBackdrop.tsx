import {composePressableStyles} from 'bases/core/utils/composePressableStyles';
import React, {useMemo} from 'react';
import {Modal as RNModal, ModalProps, Pressable, PressableProps, StyleSheet, ViewProps} from 'react-native';
import Reanimated, {WithTimingConfig} from 'react-native-reanimated';

import {usePickerBackdrop} from './usePickerBackdrop';

export const DEFAULT_BACKGROUND_COLOR = 'black';
export const DEFAULT_OPACITY = 0.4;
export const DEFAULT_FADE_IN_DURATION = 300;
export const DEFAULT_FADE_OUT_DURATION = 150;

// React Native ReanimatedのLayout Animationsを使用すると↓の不具合が発生するため、指定できないようにしています。
// https://github.com/software-mansion/react-native-reanimated/issues/2906
export type PickerBackdropProps = Omit<Reanimated.AnimateProps<ViewProps>, 'exiting' | 'entering'> & {
  isVisible: boolean;
  onPress?: () => unknown;
  afterFadeIn?: (finished?: boolean) => unknown;
  afterFadeOut?: (finished?: boolean) => unknown;
  fadeInDuration?: number;
  fadeOutDuration?: number;
  opacity?: number;
  pressableProps?: PressableProps;
  modalProps?: ModalProps;
  fadeInConfig?: WithTimingConfig;
  fadeOutConfig?: WithTimingConfig;
};

export const PickerBackdrop: React.FC<React.PropsWithChildren<PickerBackdropProps>> = ({
  isVisible,
  opacity = DEFAULT_OPACITY,
  onPress,
  afterFadeIn,
  afterFadeOut,
  fadeInDuration = DEFAULT_FADE_IN_DURATION,
  fadeOutDuration = DEFAULT_FADE_OUT_DURATION,
  pressableProps: {style: pressableStyle, ...pressableProps} = {},
  modalProps: {style: modalStyle, ...modalProps} = {},
  fadeInConfig,
  fadeOutConfig,
  children,
}) => {
  const {isModalVisible} = usePickerBackdrop({
    isVisible,
    opacity,
    afterFadeIn,
    afterFadeOut,
    fadeInDuration,
    fadeOutDuration,
    fadeInConfig,
    fadeOutConfig,
  });

  const composedPressableStyles = useMemo(
    () => composePressableStyles([styles.pressable, pressableStyle]),
    [pressableStyle],
  );

  return !isModalVisible ? null : (
    <RNModal
      visible
      statusBarTranslucent
      animationType="fade"
      transparent
      // 戻るボタンが押されたとき（onRequestClose）は、背景がタップされたときと同じ振る舞いになるようにしておく。
      onRequestClose={onPress}
      style={modalStyle}
      {...modalProps}>
      <Pressable onPress={onPress} style={composedPressableStyles} {...pressableProps} />
      {children}
    </RNModal>
  );
};

const styles = StyleSheet.create({
  pressable: {
    ...StyleSheet.absoluteFillObject,
    display: 'flex',
    opacity: 0.4,
    backgroundColor: DEFAULT_BACKGROUND_COLOR,
  },
  backdrop: {
    flex: 1,
  },
});
