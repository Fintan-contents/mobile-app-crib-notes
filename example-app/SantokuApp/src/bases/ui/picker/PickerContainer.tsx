import React from 'react';
import {StyleSheet, View, ViewProps} from 'react-native';
import Reanimated, {WithTimingConfig} from 'react-native-reanimated';

import {usePickerContainer} from './usePickerContainer';

export const DEFAULT_SLIDE_IN_DURATION = 300;
export const DEFAULT_SLIDE_OUT_DURATION = 300;

// React Native ReanimatedのLayout Animationsを使用すると↓の不具合が発生するため、指定できないようにしています。
// https://github.com/software-mansion/react-native-reanimated/issues/2906
export type PickerContainerProps = Omit<Reanimated.AnimateProps<ViewProps>, 'exiting' | 'entering'> & {
  isVisible: boolean;
  afterSlideIn?: (finished?: boolean) => unknown;
  afterSlideOut?: (finished?: boolean) => unknown;
  slideInDuration?: number;
  slideOutDuration?: number;
  slideInConfig?: WithTimingConfig;
  slideOutConfig?: WithTimingConfig;
};

export const PickerContainer: React.FC<React.PropsWithChildren<PickerContainerProps>> = ({
  isVisible,
  style,
  afterSlideIn,
  afterSlideOut,
  slideInDuration = DEFAULT_SLIDE_IN_DURATION,
  slideOutDuration = DEFAULT_SLIDE_OUT_DURATION,
  slideInConfig,
  slideOutConfig,
  children,
  ...animatedViewProps
}) => {
  const {isPickerVisible, transform, updateContentHeight} = usePickerContainer({
    isVisible,
    afterSlideIn,
    afterSlideOut,
    slideInDuration,
    slideOutDuration,
    slideInConfig,
    slideOutConfig,
  });
  return !isPickerVisible ? null : (
    <View style={[styles.container]} pointerEvents="box-none">
      <Reanimated.View style={[style, transform]} onLayout={updateContentHeight} {...animatedViewProps}>
        {children}
      </Reanimated.View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-end',
  },
});
