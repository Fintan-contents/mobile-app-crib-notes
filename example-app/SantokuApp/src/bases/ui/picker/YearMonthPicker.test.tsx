import {fireEvent, render, screen} from '@testing-library/react-native';
import {ApplicationError} from 'bases/core/errors/ApplicationError';
import React from 'react';
import {TextInputProps, ViewProps} from 'react-native';

import {PickerBackdropProps} from './PickerBackdrop';
import {PickerContainerProps} from './PickerContainer';
import {SelectPickerItemsProps} from './SelectPickerItems';
import {YearMonthPicker} from './YearMonthPicker';
import {YearMonthUtil} from './YearMonthUtil';

jest.doMock('react-native/Libraries/Utilities/Dimensions', () => ({
  get: jest.fn().mockReturnValue({width: 400, height: 1000}),
  set: jest.fn(),
  addEventListener: jest.fn().mockReturnValue({remove: jest.fn()}),
}));

describe('YearMonthPicker only with required props', () => {
  beforeAll(() => {
    const mockDate = new Date(2022, 3);
    jest.useFakeTimers('modern');
    jest.setSystemTime(mockDate);
  });
  afterAll(() => {
    jest.useRealTimers();
  });
  it('renders successfully if invisible', () => {
    const now = YearMonthUtil.now();
    render(
      <YearMonthPicker
        maximumYearMonth={now}
        minimumYearMonth={YearMonthUtil.addMonth(now, -60)}
        selectedValue={now}
        textInputProps={{testID: 'textInput'}}
        pickerBackdropProps={{pressableProps: {testID: 'pickerBackdrop'}}}
        pickerContainerProps={{testID: 'pickerContainer'}}
        pickerAccessoryProps={{containerProps: {testID: 'pickerAccessory'}}}
        pickerItemsContainerProps={{testID: 'pickerItemsContainer'}}
      />,
    );

    expect(screen).toMatchSnapshot('YearMonthPicker if invisible.');

    const pickerBackdrop = screen.queryByTestId('pickerBackdrop');
    const pickerContainer = screen.queryByTestId('pickerContainer');
    const pickerAccessory = screen.queryByTestId('pickerAccessory');
    const pickerItemsContainer = screen.queryByTestId('pickerItemsContainer');
    const textInput = screen.queryByTestId('textInput');
    expect(pickerBackdrop).toBeNull();
    expect(pickerContainer).toBeNull();
    expect(pickerAccessory).toBeNull();
    expect(pickerItemsContainer).toBeNull();
    expect(textInput).not.toBeNull();
  });

  it('renders successfully if visible', () => {
    const now = YearMonthUtil.now();
    render(
      <YearMonthPicker
        maximumYearMonth={now}
        minimumYearMonth={YearMonthUtil.addMonth(now, -60)}
        selectedValue={now}
        textInputProps={{testID: 'textInput'}}
        pickerBackdropProps={{pressableProps: {testID: 'pickerBackdrop'}}}
        pickerContainerProps={{testID: 'pickerContainer'}}
        pickerAccessoryProps={{containerProps: {testID: 'pickerAccessory'}}}
        pickerItemsContainerProps={{testID: 'pickerItemsContainer'}}
      />,
    );
    const pressableContainer = screen.getByTestId('pressableContainer');
    fireEvent.press(pressableContainer);

    expect(screen).toMatchSnapshot('YearMonthPicker if visible.');

    const pickerBackdrop = screen.queryByTestId('pickerBackdrop');
    const pickerContainer = screen.queryByTestId('pickerContainer');
    const pickerAccessory = screen.queryByTestId('pickerAccessory');
    const pickerItemsContainer = screen.queryByTestId('pickerItemsContainer');
    const textInput = screen.queryByTestId('textInput');
    expect(pickerBackdrop).not.toBeNull();
    expect(pickerContainer).not.toBeNull();
    expect(pickerAccessory).not.toBeNull();
    expect(pickerItemsContainer).not.toBeNull();
    expect(textInput).not.toBeNull();

    const textInputProps = textInput?.props as TextInputProps;
    expect(textInputProps.value).toBe('20224');
  });
});

describe('YearMonthPicker with various maximum and minimum yearMonth', () => {
  it('same yearMonth', () => {
    const yearMonth = {year: 2022, month: 4};
    render(<YearMonthPicker maximumYearMonth={yearMonth} minimumYearMonth={yearMonth} selectedValue={yearMonth} />);

    const pressableContainer = screen.getByTestId('pressableContainer');
    fireEvent.press(pressableContainer);

    const yearPicker = screen.getByTestId('yearPicker');
    const yearPickerProps = yearPicker.props as SelectPickerItemsProps<string>;
    expect(yearPickerProps.items).toEqual([{value: 2022, label: '2022'}]);
    const monthPicker = screen.getByTestId('monthPicker');
    const monthPickerProps = monthPicker.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps.items).toEqual([{value: 4, label: '4'}]);
  });

  it('same year and different month', () => {
    const maximumYearMonth = {year: 2022, month: 4};
    const minimumYearMonth = {year: 2022, month: 1};
    render(
      <YearMonthPicker
        maximumYearMonth={maximumYearMonth}
        minimumYearMonth={minimumYearMonth}
        selectedValue={maximumYearMonth}
      />,
    );

    const pressableContainer = screen.getByTestId('pressableContainer');
    fireEvent.press(pressableContainer);

    const yearPicker = screen.getByTestId('yearPicker');
    const yearPickerProps = yearPicker.props as SelectPickerItemsProps<string>;
    expect(yearPickerProps.items).toEqual([{value: 2022, label: '2022'}]);
    const monthPicker = screen.getByTestId('monthPicker');
    const monthPickerProps = monthPicker.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps.items).toEqual([
      {value: 1, label: '1'},
      {value: 2, label: '2'},
      {value: 3, label: '3'},
      {value: 4, label: '4'},
    ]);
  });

  it('different year', () => {
    const maximumYearMonth = {year: 2022, month: 4};
    const minimumYearMonth = {year: 2021, month: 11};
    render(
      <YearMonthPicker
        maximumYearMonth={maximumYearMonth}
        minimumYearMonth={minimumYearMonth}
        selectedValue={maximumYearMonth}
      />,
    );

    const pressableContainer = screen.getByTestId('pressableContainer');
    fireEvent.press(pressableContainer);

    const yearPicker = screen.getByTestId('yearPicker');
    const yearPickerProps = yearPicker.props as SelectPickerItemsProps<string>;
    expect(yearPickerProps.items).toEqual([
      {value: 2021, label: '2021'},
      {value: 2022, label: '2022'},
    ]);
    const monthPicker = screen.getByTestId('monthPicker');
    const monthPickerProps = monthPicker.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps.items).toEqual([
      {value: 1, label: '1'},
      {value: 2, label: '2'},
      {value: 3, label: '3'},
      {value: 4, label: '4'},
      {value: 5, label: '5'},
      {value: 6, label: '6'},
      {value: 7, label: '7'},
      {value: 8, label: '8'},
      {value: 9, label: '9'},
      {value: 10, label: '10'},
      {value: 11, label: '11'},
      {value: 12, label: '12'},
    ]);
  });

  it('maximum < minimum', () => {
    // コンポーネントマウント時のエラーログが出力されないようにモック化しておく
    jest.spyOn(console, 'error').mockImplementation();
    const maximumYearMonth = {year: 2022, month: 4};
    const minimumYearMonth = {year: 2022, month: 5};
    expect(() =>
      render(
        <YearMonthPicker
          maximumYearMonth={maximumYearMonth}
          minimumYearMonth={minimumYearMonth}
          selectedValue={maximumYearMonth}
        />,
      ),
    ).toThrow(new ApplicationError('Maximum yearMonth must be greater than or equal to minimum yearMonth.'));
  });
});

describe('YearMonthPicker with all props', () => {
  it('should be applied all properly', () => {
    const maximumYearMonth = {year: 2022, month: 4};
    const onSelectedItemChange = jest.fn();
    const onDismiss = jest.fn();
    const onDelete = jest.fn();
    const onCancel = jest.fn();
    const onDone = jest.fn();
    const screen = render(
      <YearMonthPicker
        maximumYearMonth={maximumYearMonth}
        minimumYearMonth={YearMonthUtil.addMonth(maximumYearMonth, -60)}
        yearSuffixLabel="年"
        monthSuffixLabel="月"
        selectedValue={maximumYearMonth}
        onSelectedItemChange={onSelectedItemChange}
        onDismiss={onDismiss}
        placeholder="please select..."
        textInputProps={{style: {color: 'red'}, testID: 'textInput'}}
        pickerItemsContainerProps={{pointerEvents: 'none', testID: 'pickerItemsContainer'}}
        pickerProps={{numberOfLines: 5}}
        pickerBackdropProps={{
          modalProps: {testID: 'pickerBackdropModal'},
          pressableProps: {testID: 'pickerBackdropPressable', style: {backgroundColor: 'green', borderColor: 'red'}},
        }}
        pickerContainerProps={{style: {backgroundColor: 'yellow', borderColor: 'black'}, testID: 'pickerContainer'}}
        pickerAccessoryProps={{
          containerProps: {style: {backgroundColor: 'blue'}, testID: 'pickerAccessory'},
          deleteLabel: 'delete',
          cancelLabel: 'cancel',
          doneLabel: 'done',
          deleteTouchableContainerProps: {testID: 'deleteLink'},
          cancelTouchableContainerProps: {testID: 'cancelLink'},
          doneTouchableContainerProps: {testID: 'doneLink'},
        }}
        onDelete={onDelete}
        onCancel={onCancel}
        onDone={onDone}
        itemColor="yellow"
        itemFontFamily="Roboto"
      />,
    );
    const pressableContainer = screen.getByTestId('pressableContainer');
    fireEvent.press(pressableContainer);

    expect(screen).toMatchSnapshot('YearMonthPicker all properly.');

    // assert pickerContainer
    const pickerContainer = screen.getByTestId('pickerContainer');
    const pickerContainerProps = pickerContainer.props as PickerContainerProps;
    expect(pickerContainerProps.style).toEqual({
      backgroundColor: 'yellow',
      borderColor: 'black',
      transform: [{translateY: 1000}],
    });

    // assert pickerItemsContainer
    const pickerItemsContainer = screen.getByTestId('pickerItemsContainer');
    const pickerItemsContainerProps = pickerItemsContainer.props as ViewProps;
    expect(pickerItemsContainerProps.pointerEvents).toBe('none');

    // assert PickerItems
    const yearPicker = screen.getByTestId('yearPicker');
    const yearPickerProps = yearPicker.props as SelectPickerItemsProps<string>;
    expect(yearPickerProps.numberOfLines).toBe(5);
    // itemsは、Itemの各Propを検証できなかったため、件数のみ検証する
    expect(yearPickerProps.items.length).toBe(6);
    const monthPicker = screen.getByTestId('monthPicker');
    const monthPickerProps = monthPicker.props as SelectPickerItemsProps<string>;
    expect(monthPickerProps.numberOfLines).toBe(5);
    // itemsは、Itemの各Propを検証できなかったため、件数のみ検証する
    expect(monthPickerProps.items.length).toBe(12);
    const yearSuffix = screen.queryByText('年');
    const monthSuffix = screen.queryByText('月');
    expect(yearSuffix).not.toBeNull();
    expect(monthSuffix).not.toBeNull();

    // assert textInput
    const textInput = screen.getByTestId('textInput');
    const textInputProps = textInput.props as TextInputProps;
    expect(textInputProps.style).toEqual({color: 'red'});
    expect(textInputProps.value).toBe('2022年4月');

    // assert pickerBackdrop
    const pickerBackdropPressable = screen.getByTestId('pickerBackdropPressable');
    const pickerBackdropModal = screen.getByTestId('pickerBackdropModal');
    const pickerBackdropProps = pickerBackdropPressable.props as PickerBackdropProps;
    fireEvent(pickerBackdropModal, 'onRequestClose');
    expect(pickerBackdropProps.style).toEqual({
      backgroundColor: 'green',
      borderColor: 'red',
      opacity: 0.4,
      bottom: 0,
      display: 'flex',
      left: 0,
      position: 'absolute',
      right: 0,
      top: 0,
    });
    expect(onDismiss).toHaveBeenCalledTimes(1);

    fireEvent.press(pressableContainer);

    // assert pickerAccessory
    const pickerAccessory = screen.getByTestId('pickerAccessory');
    const pickerAccessoryProps = pickerAccessory.props as ViewProps;
    expect(pickerAccessoryProps.style).toEqual({
      backgroundColor: 'blue',
      flexDirection: 'row',
      justifyContent: 'flex-end',
      paddingHorizontal: 10,
      paddingVertical: 8,
    });
    const deleteLink = screen.getByTestId('deleteLink');
    const deleteText = screen.queryByText('delete');
    expect(deleteText).not.toBeNull();
    fireEvent.press(deleteLink);
    expect(onDelete).toHaveBeenCalledTimes(1);
    fireEvent.press(pressableContainer);
    const cancelLink = screen.getByTestId('cancelLink');
    const cancelText = screen.queryByText('cancel');
    expect(cancelText).not.toBeNull();
    fireEvent.press(cancelLink);
    expect(onCancel).toHaveBeenCalledTimes(1);
    fireEvent.press(pressableContainer);
    const doneLink = screen.getByTestId('doneLink');
    const doneText = screen.queryByText('done');
    expect(doneText).not.toBeNull();
    fireEvent.press(doneLink);
    expect(onDone).toHaveBeenCalledTimes(1);
  });
});
