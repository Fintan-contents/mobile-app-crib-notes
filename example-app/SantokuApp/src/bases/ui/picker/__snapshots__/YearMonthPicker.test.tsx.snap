// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearMonthPicker only with required props renders successfully if invisible: YearMonthPicker if invisible. 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="pressableContainer"
>
  <View
    pointerEvents="box-only"
  >
    <TextInput
      editable={false}
      testID="textInput"
      value="20224"
    />
  </View>
</View>
`;

exports[`YearMonthPicker only with required props renders successfully if visible: YearMonthPicker if visible. 1`] = `
Array [
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    transparent={true}
    visible={true}
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "display": "flex",
          "left": 0,
          "opacity": 0.4,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="pickerBackdrop"
    />
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <View
        animatedStyle={
          Object {
            "value": Object {
              "transform": Array [
                Object {
                  "translateY": 1000,
                },
              ],
            },
          }
        }
        collapsable={false}
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "transform": Array [
              Object {
                "translateY": 1000,
              },
            ],
          }
        }
        testID="pickerContainer"
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "flex-end",
              "paddingHorizontal": 10,
              "paddingVertical": 8,
            }
          }
          testID="pickerAccessory"
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "paddingHorizontal": 10,
            }
          }
          testID="pickerItemsContainer"
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View>
              <RNCPicker
                items={
                  Array [
                    Object {
                      "label": "2017",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 2017,
                    },
                    Object {
                      "label": "2018",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 2018,
                    },
                    Object {
                      "label": "2019",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 2019,
                    },
                    Object {
                      "label": "2020",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 2020,
                    },
                    Object {
                      "label": "2021",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 2021,
                    },
                    Object {
                      "label": "2022",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 2022,
                    },
                  ]
                }
                numberOfLines={1}
                onChange={[Function]}
                selectedIndex={5}
                style={
                  Array [
                    Object {
                      "height": 216,
                    },
                    undefined,
                  ]
                }
                testID="yearPicker"
              />
            </View>
          </View>
          <Text />
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View>
              <RNCPicker
                items={
                  Array [
                    Object {
                      "label": "1",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 1,
                    },
                    Object {
                      "label": "2",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 2,
                    },
                    Object {
                      "label": "3",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 3,
                    },
                    Object {
                      "label": "4",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 4,
                    },
                    Object {
                      "label": "5",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 5,
                    },
                    Object {
                      "label": "6",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 6,
                    },
                    Object {
                      "label": "7",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 7,
                    },
                    Object {
                      "label": "8",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 8,
                    },
                    Object {
                      "label": "9",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 9,
                    },
                    Object {
                      "label": "10",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 10,
                    },
                    Object {
                      "label": "11",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 11,
                    },
                    Object {
                      "label": "12",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": 12,
                    },
                  ]
                }
                numberOfLines={1}
                onChange={[Function]}
                selectedIndex={3}
                style={
                  Array [
                    Object {
                      "height": 216,
                    },
                    undefined,
                  ]
                }
                testID="monthPicker"
              />
            </View>
          </View>
          <Text />
        </View>
      </View>
    </View>
  </Modal>,
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="pressableContainer"
  >
    <View
      pointerEvents="box-only"
    >
      <TextInput
        editable={false}
        testID="textInput"
        value="20224"
      />
    </View>
  </View>,
]
`;

exports[`YearMonthPicker with all props should be applied all properly: YearMonthPicker all properly. 1`] = `
Array [
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    testID="pickerBackdropModal"
    transparent={true}
    visible={true}
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "green",
          "borderColor": "red",
          "bottom": 0,
          "display": "flex",
          "left": 0,
          "opacity": 0.4,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="pickerBackdropPressable"
    />
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "flex": 1,
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <View
        animatedStyle={
          Object {
            "value": Object {
              "transform": Array [
                Object {
                  "translateY": 1000,
                },
              ],
            },
          }
        }
        collapsable={false}
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "yellow",
            "borderColor": "black",
            "transform": Array [
              Object {
                "translateY": 1000,
              },
            ],
          }
        }
        testID="pickerContainer"
      >
        <View
          style={
            Object {
              "backgroundColor": "blue",
              "flexDirection": "row",
              "justifyContent": "flex-end",
              "paddingHorizontal": 10,
              "paddingVertical": 8,
            }
          }
          testID="pickerAccessory"
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="deleteLink"
          >
            <Text
              style={
                Object {
                  "color": "#d9545e",
                  "fontWeight": "bold",
                  "paddingHorizontal": 10,
                }
              }
            >
              delete
            </Text>
          </View>
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          />
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="cancelLink"
          >
            <Text
              style={
                Object {
                  "color": "#4577CC",
                  "fontWeight": "bold",
                  "paddingHorizontal": 10,
                }
              }
            >
              cancel
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="doneLink"
          >
            <Text
              style={
                Object {
                  "color": "#4577CC",
                  "fontWeight": "bold",
                  "paddingHorizontal": 10,
                }
              }
            >
              done
            </Text>
          </View>
        </View>
        <View
          pointerEvents="none"
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "paddingHorizontal": 10,
            }
          }
          testID="pickerItemsContainer"
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View>
              <RNCPicker
                items={
                  Array [
                    Object {
                      "label": "2017",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 2017,
                    },
                    Object {
                      "label": "2018",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 2018,
                    },
                    Object {
                      "label": "2019",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 2019,
                    },
                    Object {
                      "label": "2020",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 2020,
                    },
                    Object {
                      "label": "2021",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 2021,
                    },
                    Object {
                      "label": "2022",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 2022,
                    },
                  ]
                }
                numberOfLines={5}
                onChange={[Function]}
                selectedIndex={5}
                style={
                  Array [
                    Object {
                      "height": 216,
                    },
                    undefined,
                  ]
                }
                testID="yearPicker"
              />
            </View>
          </View>
          <Text>
            年
          </Text>
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View>
              <RNCPicker
                items={
                  Array [
                    Object {
                      "label": "1",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 1,
                    },
                    Object {
                      "label": "2",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 2,
                    },
                    Object {
                      "label": "3",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 3,
                    },
                    Object {
                      "label": "4",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 4,
                    },
                    Object {
                      "label": "5",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 5,
                    },
                    Object {
                      "label": "6",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 6,
                    },
                    Object {
                      "label": "7",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 7,
                    },
                    Object {
                      "label": "8",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 8,
                    },
                    Object {
                      "label": "9",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 9,
                    },
                    Object {
                      "label": "10",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 10,
                    },
                    Object {
                      "label": "11",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 11,
                    },
                    Object {
                      "label": "12",
                      "testID": undefined,
                      "textColor": 4294967040,
                      "value": 12,
                    },
                  ]
                }
                numberOfLines={5}
                onChange={[Function]}
                selectedIndex={3}
                style={
                  Array [
                    Object {
                      "height": 216,
                    },
                    undefined,
                  ]
                }
                testID="monthPicker"
              />
            </View>
          </View>
          <Text>
            月
          </Text>
        </View>
      </View>
    </View>
  </Modal>,
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="pressableContainer"
  >
    <View
      pointerEvents="box-only"
    >
      <TextInput
        editable={false}
        placeholder="please select..."
        style={
          Object {
            "color": "red",
          }
        }
        testID="textInput"
        value="2022年4月"
      />
    </View>
  </View>,
]
`;
